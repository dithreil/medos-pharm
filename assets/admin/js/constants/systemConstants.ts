import {IPagination} from "../interfaces/request-params";

const selectPagination: IPagination = {
    descending: false,
    limit: 10,
    page: 1,
    rowsNumber: 10,
    rowsPerPage: null,
    sortBy: null,
};
export default {
    selectPagination,
}
