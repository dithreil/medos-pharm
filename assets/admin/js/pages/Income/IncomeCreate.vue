<template>
  <div>
    <q-breadcrumbs class="q-mb-xl employee-details__breadcrumbs">
      <q-breadcrumbs-el icon="home">
        <a class="q-breadcrumbs__el q-link" href="/admin">Панель Администратора</a>
      </q-breadcrumbs-el>
      <q-breadcrumbs-el
          icon="widgets"
          label="Поступления товаров"
          :to="{name: 'IncomeList'}"
      />
      <q-breadcrumbs-el
          icon="face"
          label="Создание поступления товаров"
      />
    </q-breadcrumbs>
    <q-card class="emploee-edit">
      <q-form
          class="emploee-edit__form col w-100"
          ref="form"
          @submit="saveOnly"
      >
        <q-card-section>
          <div class="text-h6">
            Создать поступление товаров
          </div>
        </q-card-section>
        <q-card-section>
          <div class="row">
            <div class="col-4" style="padding: 0 30px">
              <q-input filled v-model="model.date">
                <template v-slot:append>
                  <q-icon name="event" class="cursor-pointer">
                    <q-popup-proxy transition-show="scale" transition-hide="scale">
                      <q-date v-model="model.date" mask="DD.MM.YYYY HH:mm:SS">
                        <div class="row items-center justify-end">
                          <q-btn v-close-popup label="Close" color="primary" flat />
                        </div>
                      </q-date>
                    </q-popup-proxy>
                  </q-icon>
                  <q-icon name="access_time" class="cursor-pointer">
                    <q-popup-proxy transition-show="scale" transition-hide="scale">
                      <q-time v-model="model.date" mask="DD.MM.YYYY HH:mm:SS" format24h>
                        <div class="row items-center justify-end">
                          <q-btn v-close-popup label="Close" color="primary" flat />
                        </div>
                      </q-time>
                    </q-popup-proxy>
                  </q-icon>
                </template>
              </q-input>
            </div>
            <div class="col-4" style="padding: 0 30px">
              <q-select
                  v-model="model.store"
                  option-label="name"
                  emit-value
                  :options="storeTableData"
                  label="Торговая точка"
                  use-input
                  filled
                  @filter="fetchStores"
                  lazy-rules
                  :rules="[
                                    (val) => !!val || $errorMessages.REQUIRED,
                                ]"
                  behavior="menu"
              >
                <template v-slot:option="scope">
                  <q-item  v-bind="scope.itemProps"
                           v-on="scope.itemEvents"
                  >
                    <q-item-section>
                      <q-item-label v-html="scope.opt.name" />
                      <q-item-label caption>{{ scope.opt.address }}</q-item-label>
                    </q-item-section>
                  </q-item>
                </template>
                <template v-slot:no-option>
                  <q-item>
                    <q-item-section class="text-grey">
                      Нет результата
                    </q-item-section>
                  </q-item>
                </template>
              </q-select>
            </div>
            <div class="col-4" style="padding: 0 30px">
              <q-select
                  v-model="model.supplier"
                  option-label="name"
                  emit-value
                  :options="supplierTableData"
                  label="Поставщик"
                  use-input
                  filled
                  @filter="fetchSuppliers"
                  lazy-rules
                  :rules="[
                                    (val) => !!val || $errorMessages.REQUIRED,
                                ]"
                  behavior="menu"
              >
                <template v-slot:option="scope">
                  <q-item  v-bind="scope.itemProps"
                           v-on="scope.itemEvents"
                  >
                    <q-item-section>
                      <q-item-label v-html="scope.opt.name" />
                      <q-item-label caption>{{ scope.opt.address }}</q-item-label>
                    </q-item-section>
                  </q-item>
                </template>
                <template v-slot:no-option>
                  <q-item>
                    <q-item-section class="text-grey">
                      Нет результата
                    </q-item-section>
                  </q-item>
                </template>
              </q-select>
            </div>
          </div>
          <DocumentTable :model="model" />
        </q-card-section>
        <q-card-actions align="left">
          <q-btn
              unelevated
              label="Сохранить"
              color="primary"
              type="submit"
          />
          <q-btn
              unelevated
              label="Провести"
              color="primary"
              type="submit"
              @click="saveAndPost"
          />
        </q-card-actions>
      </q-form>
    </q-card>
  </div>
</template>

<script lang="ts">
import {mapActions, mapGetters} from 'vuex';
import * as validationHelpers from '../../validation/helpers';
import {Component, Ref, Vue} from 'vue-property-decorator';
import {IDocumentIncomeCreateEditData} from '../../interfaces/income';
import {QForm} from 'quasar';
import {IRequestParams} from '../../interfaces/request-params';
import {IStoreData} from '../../interfaces/store';
import {ISupplierData} from '../../interfaces/supplier';
import DocumentTable from '../../components/Document/DocumentTable.vue';
import moment from 'moment';
import IncomeDocument from '../../models/IncomeDocument';
import {error} from '../../utils/notifizer';


@Component({
    components: {DocumentTable},
    computed: {
        ...mapGetters({
            suppliersData: 'supplier/suppliersData',
            storesData: 'store/storesData',
            // user: 'user/userData',
        }),
    },
    methods: {
        ...mapActions({
            postIncome: 'income/postIncome',
            createIncome: 'income/createIncome',
            updateStoreRequestParams: 'store/updateStoreRequestParams',
            updateSupplierRequestParams: 'supplier/updateSupplierRequestParams',
        }),
        ...validationHelpers,
    },
})
export default class IncomeCreate extends Vue {
  @Ref('form') readonly form!: QForm;

  protected updateStoreRequestParams!: (payload: IRequestParams) => any;
  protected updateSupplierRequestParams!: (payload: IRequestParams) => any;
  protected createIncome!: ({payload}: {payload: IDocumentIncomeCreateEditData}) => any;
  protected postIncome!: ({payload}: {payload: {action: string}, id: string}) => any;


  protected storesData!: IStoreData
  protected suppliersData!: ISupplierData

  protected model: IncomeDocument = new IncomeDocument();

  get storeTableData() {
      return this.storesData?.items || [];
  };

  get supplierTableData() {
      return this.suppliersData?.items || [];
  };


  mounted() {
      this.model.date = moment().format('DD.MM.YYYY HH:mm:SS');
  }

  isFormInvalid() {
      return this.form.validate();
  };

  fetchStores(val : string, update : () => any) {
      this.updateStoreRequestParams({...this.$constants.systemConstants.selectPagination, filter: val})
          .finally(() => {
              update();
          });
  }

  fetchSuppliers(val : string, update : () => any) {
      this.updateSupplierRequestParams({...this.$constants.systemConstants.selectPagination, filter: val})
          .finally(() => {
              update();
          });
  }
  async saveDocument() {
      const isValid = await this.model.isValid();

      if (!isValid) {
          error('Данные формы некорректны!');

          return null;
      }

      return await this.createIncome({payload: this.model.getDataForServer()});
  };

  async saveOnly() {
      const response = await this.saveDocument();

      if (response && 201 === response.status) {
          await this.$router.push({name: 'IncomeEdit', params: {id: response.data.id}});
      }
  }

  async saveAndPost() {
      const response = await this.saveDocument();

      if (response && 201 === response.status) {
          const postResponse = await this.postIncome({id: response.data.id, payload: {action: 'income.enter'}});
          if (204 === postResponse.status) {
              await this.$router.push({name: 'IncomeEdit', params: {id: response.data.id}});
          }
      }
  }
};
</script>
