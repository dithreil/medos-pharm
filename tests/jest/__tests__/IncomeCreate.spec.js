import {mountQuasar} from '@quasar/quasar-app-extension-testing-unit-jest';
import IncomeCreate from '../../../assets/admin/js/pages/Income/IncomeCreate.vue';
import Vuex from 'vuex';
import {createLocalVue} from '@vue/test-utils';
import {
    QBreadcrumbs,
    QBreadcrumbsEl,
    QBtn,
    QCard,
    QCardActions,
    QCardSection,
    QForm,
    QInput,
    QSelect,
    QTable,
} from 'quasar';
import systemConstants from '../../../assets/admin/js/constants/systemConstants';
const localVue = createLocalVue();
localVue.use(Vuex);

describe('IncomeCreate', () => {
    let wrapper;
    let store;
    let actions;
    function mountComponent(store) {
        wrapper = mountQuasar(IncomeCreate, {
            quasar: {
                components: {
                    QTable,
                    QBtn,
                    QBreadcrumbs,
                    QBreadcrumbsEl,
                    QCard,
                    QForm,
                    QCardSection,
                    QInput,
                    QSelect,
                    QCardActions,
                },
            },
            mount: {
                store,
                localVue,
            },
        });
    }
    beforeEach(async()=> {
        actions = {
            'income/postIncome': jest.fn(),
            'income/createIncome': jest.fn(),
            'store/updateStoreRequestParams': jest.fn(),
            'supplier/updateSupplierRequestParams': jest.fn(),
        };
        store = new Vuex.Store({
            actions,
            getters: {
                'supplier/suppliersData': () => null,
                'store/storesData': () => null,
            },
        });
        mountComponent(store);
    });

    test('matches snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    });
});
