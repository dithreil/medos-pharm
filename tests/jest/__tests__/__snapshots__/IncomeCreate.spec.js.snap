// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomeCreate matches snapshot 1`] = `
<div>
  <q-breadcrumbs-stub
    activecolor="primary"
    class="q-mb-xl employee-details__breadcrumbs"
    gutter="sm"
    separator="/"
  >
    <q-breadcrumbs-el-stub
      icon="home"
    >
      <a
        class="q-breadcrumbs__el q-link"
        href="/admin"
      >
        Панель Администратора
      </a>
    </q-breadcrumbs-el-stub>
     
    <q-breadcrumbs-el-stub
      icon="widgets"
      label="Поступления товаров"
      to="[object Object]"
    />
     
    <q-breadcrumbs-el-stub
      icon="face"
      label="Создание поступления товаров"
    />
  </q-breadcrumbs-stub>
   
  <q-card-stub
    class="emploee-edit"
    tag="div"
  >
    <q-form-stub
      class="emploee-edit__form col w-100"
    >
      <q-card-section-stub
        tag="div"
      >
        <div
          class="text-h6"
        >
          
          Создать поступление товаров
        
        </div>
      </q-card-section-stub>
       
      <q-card-section-stub
        tag="div"
      >
        <div
          class="row"
        >
          <div
            class="col-4"
            style="padding: 0px 30px;"
          >
            <q-input-stub
              filled="true"
              type="text"
              value=""
            />
          </div>
           
          <div
            class="col-4"
            style="padding: 0px 30px;"
          >
            <q-select-stub
              behavior="menu"
              emitvalue="true"
              filled="true"
              inputdebounce="500"
              label="Торговая точка"
              lazyrules="true"
              optionlabel="name"
              options=""
              rules="[Function]"
              tabindex="0"
              useinput="true"
              value="[object Object]"
              virtualscrollsliceratioafter="1"
              virtualscrollsliceratiobefore="1"
              virtualscrollstickysizeend="0"
              virtualscrollstickysizestart="0"
            />
          </div>
           
          <div
            class="col-4"
            style="padding: 0px 30px;"
          >
            <q-select-stub
              behavior="menu"
              emitvalue="true"
              filled="true"
              inputdebounce="500"
              label="Поставщик"
              lazyrules="true"
              optionlabel="name"
              options=""
              rules="[Function]"
              tabindex="0"
              useinput="true"
              value="[object Object]"
              virtualscrollsliceratioafter="1"
              virtualscrollsliceratiobefore="1"
              virtualscrollstickysizeend="0"
              virtualscrollstickysizestart="0"
            />
          </div>
        </div>
         
        <documenttable-stub
          model="[object Object]"
        />
      </q-card-section-stub>
       
      <q-card-actions-stub
        align="left"
      >
        <q-btn-stub
          align="center"
          color="primary"
          label="Сохранить"
          ripple="true"
          type="submit"
          unelevated="true"
        />
         
        <q-btn-stub
          align="center"
          color="primary"
          label="Провести"
          ripple="true"
          type="submit"
          unelevated="true"
        />
      </q-card-actions-stub>
    </q-form-stub>
  </q-card-stub>
</div>
`;
